### ISSUE TITLE
Behavior during reconnect

### ISSUE THREAD

**Issue by snapbug** [aembke/fred.rs#145]
Wanted to check some behavior we're seeing when we have a node in our cluster fails over, or loses connection.

For context, we run what is essentially a proxy service for many different storage backends, taking requests for items, modifying and routing them to the correct backend.

We see three things we wanted to check --

1. Connections are re-established to all nodes, not just the failed node
We see this manifest with logs (from the [`on_error` channel](https://docs.rs/fred/latest/fred/interfaces/trait.ClientLike.html#method.on_error)) for each node of the form `Ending replica reader task from xxx.xxx.xxx.xxx:6379 due to None`, with the original first error being related to the first connection dropping.

2. During a connection failover we see the memory usage of the containers for our service increase over time, often with two phases (highlighted example).
<img width="628" alt="image" src="https://github.com/aembke/fred.rs/assets/945966/98fed545-2e18-462a-a29e-1b0a3044c95b">
<img width="617" alt="240226962-dc212736-630c-4e63-b569-fd805bd01d86" src="https://github.com/aembke/fred.rs/assets/945966/23d61a16-87e2-46c3-be8d-b8d6ddc483ad">

3. During these periods of connection issues we see every request hitting our internal timeouts (using tokio's `timeout` with a setting about 20ms higher than fred's `PerformanceConfig.default_command_timeout_ms`), e.g. ->
<img width="606" alt="image" src="https://github.com/aembke/fred.rs/assets/945966/82e16734-da34-4dd4-a916-09dadbfbaf80">

(The yellow and purple lines are for the same underlying redis cluster, using amazon's [data-tiering cache product](https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/data-tiering.html), while the blue is a regular memory redis (also via AWS elasticache).

The settings that we're using are mostly defaulted...

```rust
   let mut redis_config = RedisConfig::from_url(&format!("redis-cluster://{}", host))     
       .expect("Could not create `RedisConfig`");                                         
   redis_config.replica = fred::types::ReplicaConfig {                                              
       lazy_connections: false,                                                                     
       connection_error_count: 1,                                                                   
       ..Default::default()                                                                         
   };                                                                                               
   let performance_config = fred::types::PerformanceConfig {                                        
       default_command_timeout_ms: 50,                                                             
       max_command_attempts: 1,                                                                     
       backpressure: fred::types::BackpressureConfig {                                              
           disable_auto_backpressure: true,                                                         
           ..Default::default()                                                                     
       },                                                                                           
       ..Default::default()                                                                         
   };                                                                                               
   let reconnect_policy = fred::types::ReconnectPolicy::default();                                  
```

Any guidance for whether this is expected? And if it is then any settings that we've missed that could help alleviate these observed problems would be really helpful.

Can provide more logs/metrics that you think might be useful. TIA 🙇

**Comment by aembke** [comment]
Hey @snapbug, thanks for writing all this up, this is really helpful.

I have a few follow up questions on some of your charts, but I'll go over your list in order. In general I think they show at least one bug in the client, but the memory chart could maybe be explained by something else. 

1. Yeah, that's expected if you set `lazy_connections: false`. There are a few improvements I'd like to make to the replica connection management logic, and this is one of them. Right now callers have two choices via the `lazy_connections` flag w/r/t reconnection events: always disconnect and reconnect to everything, or always drop everything and reconnect lazily. A better middle ground might be one that holds open a connection to each replica, but reconnects [only as needed](https://github.com/aembke/fred.rs/blob/909c1099c074f69b6e5f63a602fb35b9b2f3d1a6/src/router/clustered.rs#L179). I would expect the lazy policy to be quite a bit quicker in some cases though. 

2. This one seems really problematic.
  * Can you describe the failover scenario? Were all the nodes (primary and replica) off at once? Did the cluster fully recover at any point in those charts after ~15:50?
  * Does your app layer implement any form of backpressure? 
  * The memory growth can and probably does indicate a bug here, but depending on your use case or intended configuration it could also be expected. For example, if you want the client to queue up commands in memory while it waits on the cluster to fully recover then you might also expect to see a similar memory growth if the cluster is down for relatively long periods of time. When combined with 3 though I think it points to at least one bug. 
* You might be able to work around this by setting `lazy_connections: true` and increasing `max_command_attempts` to something greater than `connection_error_count`. It probably won't fix everything, but those are the main differences between your config and the one in the tests.

3. There's at least one bug here if the expected 50ms timeout doesn't run. 

I'll try to come up with some test scenarios to repro this with docker, but I might need some more info on how exactly to best repro both of the issues you saw here. In the meantime I would say maybe try adding the `check-unresponsive` feature flag and tweak few of the settings mentioned above.

**Comment by snapbug** [comment]
> Hey @snapbug, thanks for writing all this up, this is really helpful.
> 
> I have a few follow up questions on some of your charts, but I'll go over your list in order. In general I think they show at least one bug in the client, but the memory chart could maybe be explained by something else.
> 
> 1. Yeah, that's expected if you set `lazy_connections: false`. There are a few improvements I'd like to make to the replica connection management logic, and this is one of them. Right now callers have two choices via the `lazy_connections` flag w/r/t reconnection events: always disconnect and reconnect to everything, or always drop everything and reconnect lazily. A better middle ground might be one that holds open a connection to each replica, but reconnects [only as needed](https://github.com/aembke/fred.rs/blob/909c1099c074f69b6e5f63a602fb35b9b2f3d1a6/src/router/clustered.rs#L179). I would expect the lazy policy to be quite a bit quicker in some cases though.

Ok, we can try with `lazy_connections: true` now that the replica discovery has been fixed. I think our expected outcome was that when a single primary node goes down, and is replaced, then we'd only see connections involving that shard be cycled in any manner, rather than the entire pool. But happy to see that this is at least somewhat expected 🙂 

> 2. This one seems really problematic.
> 
> * Can you describe the failover scenario? Were all the nodes (primary and replica) off at once? Did the cluster fully recover at any point in those charts after ~15:50?

We've been able to reproduce the memory graphs by manually failing over a primary node, while the replica is still available to serve. In reproductions we've seen that sometimes the container will sit at an elevated (but not OOM'd) level of memory, and can still serve responses. Our typical node recovery period in this cluster is about 45minutes.

> * Does your app layer implement any form of backpressure?

Kinda, sorta, we give up pretty quickly if we notice problems with any of our backends. But nothing super strict/official. We wrap all fetching from the different backends with a `tokio::timeout`, which has a fairly low timeout (but higher than the `default_command_timeout_ms` setting we chose).

In the past we've had slower backends, so I don't think the request queueing up while we wait for the wrapped timeout is going to be a source of issue, but we can try wrapping with a closer to `default_command_timeout_ms` overall timeout.

> * The memory growth can and probably does indicate a bug here, but depending on your use case or intended configuration it could also be expected. For example, if you want the client to queue up commands in memory while it waits on the cluster to fully recover then you might also expect to see a similar memory growth if the cluster is down for relatively long periods of time. When combined with 3 though I think it points to at least one bug.

We're not going to wait, and return missing/unavailable to our clients (wrapping in a large `tokio::timeout`)

> * You might be able to work around this by setting `lazy_connections: true` and increasing `max_command_attempts` to something greater than `connection_error_count`. It probably won't fix everything, but those are the main differences between your config and the one in the tests.

Great, we can certainly try that for our next iteration of testing 🙇 

> 
> 3. There's at least one bug here if the expected 50ms timeout doesn't run.
> 
> I'll try to come up with some test scenarios to repro this with docker, but I might need some more info on how exactly to best repro both of the issues you saw here. In the meantime I would say maybe try adding the `check-unresponsive` feature flag and tweak few of the settings mentioned above.

Ah, I should have mentioned the feature flags we have turned on 🤦 ... we've got the following features on:
```toml
    "reconnect-on-auth-error",
    "check-unresponsive",
    "pool-prefer-active",
    "enable-rustls",
    "no-client-setname",
    "replicas",
```

As for repro - we've been using manually triggered failovers of primary nodes in the cluster, and for a few minutes before during and after pushing through our peak traffic (~17k qps per proxy node).

@2timesjay can help with any follow ups from suggestions you've given, as I'm going to be afk for a couple of weeks, thanks for all the help 🙇

**Comment by aembke** [comment]
Sounds good @snapbug. 

Also, @DDtKey brings up a good point - depending on how you're failing over these nodes the `default_connection_timeout_ms` may be delaying things quite a bit, resulting in those 70ms timeouts. 

A few other ideas:
* The `network_timeout` field from the `check-unresponsive` feature may have the same effect. 
* The reconnect policy there also uses a constant 1 second delay, so even one failed reconnect attempt will result in many timed out commands. 

That might fix several issues here, but I'm still concerned about the 50ms timeout not running in your example.

Here's a more detailed config example to try. This is closer to how I set things up in scale tests:

```rust
use fred::prelude::*;
use fred::globals;

async fn example() -> Result<(), RedisError> {
  // usually roughly matches `network_timeout_ms`
  globals::set_default_connection_timeout_ms(4000);
  globals::set_unresponsive_interval_ms(500);

  let config = RedisConfig {
    replica: ReplicaConfig {
      lazy_connections: true,
      ignore_reonnection_errors: true,
      connection_error_count: 1,
      primary_fallback: true,
      Default::default()
    },
    .. Default::default()
  };
  let perf = PerformanceConfig {
    auto_pipeline: true,
    max_command_attempts: 3,
    backpressure: BackpressureConfig {
      max_in_flight_commands: 5000,
      policy: BackpressurePolicy::Drain,
      .. Default::default()
    },
    // this is approximately how long it takes my local cluster to detect a downed node and promote a replica. if your intention is that write requests try again at least once, then you may need to increase this
    default_command_timeout_ms: 5000,
    // this should usually be less than the default command timeout above if you want commands to have time for a second attempt
    #[cfg(feature = "check-unresponsive")]
    network_timeout_ms: 4000,
    .. Default::default()
  };
  // exponential backoff with these settings will likely work better than a constant 1sec delay
  let policy = ReconnectPolicy::new_exponential(0, 50, 30_000, 2);
  // ...
}
```

One more question:

> We've been able to reproduce the memory graphs by manually failing over a primary node, while the replica is still available to serve. In reproductions we've seen that sometimes the container will sit at an elevated (but not OOM'd) level of memory, and can still serve responses. Our typical node recovery period in this cluster is about 45minutes.

I'm going to try to repro this with `bollard` in a module in `bin/replica_test`. Just to double check my understanding:
1. A primary node shuts down, but not the replica (sounds like there's one replica?).
2. The cluster recognizes the primary failing and promotes the replica to primary.
3. Some time later (maybe ~45 min) the old primary recovers and rejoins the cluster.

Is that right? Or was there more of a delay between 1 and 2? 

@DDtKey let's go over the `default_connection_timeout_ms` options in https://github.com/aembke/fred.rs/issues/144

**Comment by DDtKey** [comment]
> @DDtKey let's go over the default_connection_timeout_ms options in https://github.com/aembke/fred.rs/issues/144

Yes, sure. Thanks!

What's about this one: https://github.com/aembke/fred.rs/issues/145#issuecomment-1561532606 ? It's about command-timeout and retry logic. I just wondering if fred can support something more flexible for commands, like retries with backoff for commands or apply timeout only to the single attempt. 

Because it's quite important to be able to retry command once a connection is re-established without timeout. It's just a bit confusing to me, because usually command-level timeouts are applied only for single attempt and not for full cycle of retries 🤔

**Comment by 2timesjay** [comment]
> I'm going to try to repro this with bollard in a module in bin/replica_test. Just to double check my understanding:
> 
> A primary node shuts down, but not the replica (sounds like there's one replica?).
> The cluster recognizes the primary failing and promotes the replica to primary.
> Some time later (maybe ~45 min) the old primary recovers and rejoins the cluster.
> Is that right? Or was there more of a delay between 1 and 2?

Yes that's correct.

we're gather more data, will update you.

**Comment by aembke** [comment]
@2timesjay I made some significant changes to the replica routing logic in 7.x (https://crates.io/crates/fred/7.0.0-rc.1). Can you try against that version?

**Comment by 2timesjay** [comment]
@aembke Thanks for getting back to us. This app is in maintenance mode right now, but we'll see if we can get the chance to look it over and test it.

**Comment by aembke** [comment]
Thanks @2timesjay. In the meantime I'm gonna close this, but I'll reopen if needed.
